## TF IDF

TF-IDF (Term Frequency-Inverse Document Frequency) is a numerical statistic used in NLP to weigh the importance of a word within a document relative to a larger corpus. It highlights unique, meaningful words by multiplying how often a word appears in a document (TF) by the inverse frequency of the word across all documents (IDF).

**Term Frequency (TF)**  
Measures how frequently a word occurs in a document.  
TF = Number of timers a word t appears in document d / Total words in document d

**Inverse Term Frequency**    
Reduces the weight of common words and increases the weight of rare words.  
IDF = log( Total number of documents N / Number of documents containing word t)  
TF-IDF Formula = TF * IDF  
  
IDF will be higher if word is rare in documents. IDF will be lower if word  is coming in most of the documents.

In [1]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [12]:
import pandas as pd

In [7]:
d0 = 'The cat sits'
d1 = 'The dog runs'

string = [d0, d1]

In [9]:
tfidf = TfidfVectorizer()
result = tfidf.fit_transform(string)

In [19]:
print(tfidf.idf_)
feature_names = tfidf.get_feature_names_out()
print(feature_names)

[1.40546511 1.40546511 1.40546511 1.40546511 1.        ]
['cat' 'dog' 'runs' 'sits' 'the']


In [15]:
print('\nidf values:')
for ele1, ele2 in zip(tfidf.get_feature_names_out(), tfidf.idf_):
    print(ele1, ':', ele2)


idf values:
cat : 1.4054651081081644
dog : 1.4054651081081644
runs : 1.4054651081081644
sits : 1.4054651081081644
the : 1.0


In [20]:
df = pd.DataFrame(result.toarray(), columns=feature_names, index=["Doc 1", "Doc 2"])

In [21]:
print(df)

            cat       dog      runs      sits       the
Doc 1  0.631667  0.000000  0.000000  0.631667  0.449436
Doc 2  0.000000  0.631667  0.631667  0.000000  0.449436



| Word | Document 1 Score | Document 2 Score | Explanation |
| :--- | :--- | :--- | :--- |
| **the** | 0.449 | 0.449 | Low score because it's in every document. |
| **cat** | 0.632 | 0.000 | High score in Doc 1; it defines what the doc is about. |
| **dog** | 0.000 | 0.632 | High score in Doc 2; it identifies the subject. |
| **runs** | 0.000 | 0.632 | Unique to Doc 2. |